internal/repository/validator.go

```diff
@@ -7,198 +7,339 @@ import (
+        first_block,
+		last_block,
+		claim_hash,
+		transaction_hash,
+		status
+	FROM 
+		epoch
+	WHERE 
+		application_address=@appAddress AND status=@status
+	ORDER BY
+		index ASC`
 
 	args := pgx.NamedArgs{
-		"address": appAddress,
+		"appAddress": application,
+		"status":     EpochStatusProcessedAllInputs,
 	}
 
-	err := pg.db.QueryRow(ctx, query, args).Scan(&block)
+	rows, err := pg.db.Query(ctx, query, args)
 	if err != nil {
-		return 0, fmt.Errorf("QueryRow failed: %v\n", err)
+		return nil, fmt.Errorf("GetProcessedEpochs failed: %w\n", err)
 	}
 
-	return block, nil
+	var (
+		id, index, firstBlock, lastBlock uint64
+		appAddress                       Address
+		claimHash, transactionHash       *Hash
+		status                           string
+		results                          []*Epoch
+	)
+
+	scans := []any{
+		&id, &appAddress, &index, &firstBlock, &lastBlock, &claimHash, &transactionHash, &status,
+	}
+	_, err = pgx.ForEachRow(rows, scans, func() error {
+		epoch := &Epoch{
+			Id:              id,
+			Index:           index,
+			AppAddress:      appAddress,
+			FirstBlock:      firstBlock,
+			LastBlock:       lastBlock,
+			ClaimHash:       claimHash,
+			TransactionHash: transactionHash,
+			Status:          EpochStatus(status),
+		}
+		results = append(results, epoch)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("GetProcessedEpochs failed: %w\n", err)
+	}
+	return results, nil
 }
 
-func (pg *Database) GetAllOutputsFromProcessedInputs(
-	ctx context.Context,
-	startBlock uint64,
-	endBlock uint64,
-	timeout *time.Duration,
-) ([]Output, error) {
-	ctxTimeout, cancel := context.WithTimeout(ctx, *timeout)
-	defer cancel()
-	for {
-		select {
-		case <-ctxTimeout.Done():
-			return nil, fmt.Errorf("GetAllOutputsFromProcessedInputs timeout") // timeout
-		default:
-			outputs, err := pg.getAllOutputsFromProcessedInputs(ctxTimeout, startBlock, endBlock)
-			if outputs != nil {
-				return outputs, nil
-			}
-
-			if err != nil {
-				return nil, err
-			}
+// GetLastInputOutputHash returns the outputs Merkle tree hash calculated
+// by the Cartesi Machine after it processed the last input in the provided
+// epoch.
+func (pg *Database) GetLastInputOutputHash(ctx context.Context, epoch *Epoch) (*Hash, error) {
+	if epoch == nil {
+		return nil, NilEpoch
+	}
+
+	//Get Epoch from Database
+	epoch, err := pg.GetEpoch(ctx, epoch.Index, epoch.AppAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	//Check Epoch Status
+	switch epoch.Status { //nolint:exhaustive
+	case EpochStatusOpen, EpochStatusClosed:
+		err := fmt.Errorf(
+			"epoch '%d' of app '%v' is still being processed",
+			epoch.Index, epoch.AppAddress,
+		)
+		return nil, err
+	default:
+		break
+	}
+
+	//Get epoch last input
+	query := fmt.Sprintf(`
+	SELECT
+		outputs_hash
+	FROM 
+		input
+	WHERE
+		epoch_id = %d
+	ORDER BY
+		index DESC
+	LIMIT 1
+	`, epoch.Id)
+	var outputHash Hash
+
+	err = pg.db.QueryRow(ctx, query).Scan(&outputHash)
```

Comment:
You should use prepared statements and named args here to keep consistent with the rest of the code

Reply:
Done in 9d1b31813a6d74c82792ff5cf8a170cbb0912d14